### PR TITLE
Reworking Sorting Module for Consistency

### PR DESCRIPTION
**Pull request by CodingTil**
See Issue: https://github.com/alexfertel/rust-algorithms/issues/37

I had to revert commit https://github.com/alexfertel/rust-algorithms/commit/c9a3ea6eb13a702d1361565d3ce68a2836a794a2 as it did not successfully sort descending arrays.

I based my rework on the proposal in https://github.com/alexfertel/rust-algorithms/issues/37#issuecomment-1608191250
 - Trait for an inplace function
 - Trait for an function that returns a value
Note that inplace here does not refer to whether the actual sorting algorithm is inplace, i.e. requires no extra space (https://www.geeksforgeeks.org/sorting-terminology/), but rather do denote whether the output is stored in the same memory location of the input or not.

Let me know whether these changes follow good practice and are practical.

### CODE REVIEW

**Comment by CodingTil** [comment]
I hope one can squash commits in the PR.

File: src/sorting/bogo_sort.rs [hunk 1]
```diff
@@ -1,13 +1,17 @@
-use super::traits::MutableSorter;
 use crate::math::PCG32;
+use crate::sorting::traits::{InplaceSorter, Sorter};
 use std::time::{SystemTime, UNIX_EPOCH};
 
-const DEFAULT: u64 = 4294967296;
+const DEFAULT: u64 = u64::pow(2, 32);
```
Comment (alexfertel) [inline src/sorting/bogo_sort.rs]:
```suggestion
const DEFAULT: u64 = 2 << 32 // 2 ^ 32
```

File: src/sorting/macros.rs [hunk 2]
```diff
@@ -13,56 +11,124 @@ macro_rules! assert_not_sorted {
...
 
 macro_rules! sorting_tests {
-    ($sorter: expr, inplace) => {
-        #[test]
-        fn basic() {
-            let mut array = [5, 4, 1, 6, 0];
-            $sorter(&mut array);
-            assert_sorted!(&array);
-        }
+    ($sorter: expr, $mod_name: ident, inplace) => {
```
Comment (alexfertel) [inline src/sorting/macros.rs:14]: Great job here, I think it justifies the use of `paste`.

**Review by alexfertel — COMMENTED** [review]
Overall great job, thank you for the thorough review & contribution ❤️ 

> I had to revert commit https://github.com/alexfertel/rust-algorithms/commit/c9a3ea6eb13a702d1361565d3ce68a2836a794a2 as it did not successfully sort descending arrays.

Oh, good catch! I missed this 🚀 

> just use a Sorter Trait everywhere, and then provide two methods: sort and sort_inplace

Could you elaborate on why you didn't just use one `Sorter` trait? I was thinking something like:

```rust
pub trait Sorter<T: Ord + Copy> {
    fn sort_inplace(arr: &mut [T]);

    fn sort(arr: &[T]) -> Vec<T> {
        let mut result = arr.to_vec();
        Self::sort_inplace(&mut result);
        result
    }
}
```
---
I think I would rather have the trait implementation after the sorting function itself. That is, if you were to read top-to-bottom, you would find the algorithm first, and then the abstracting traits. This appeals better to someone new to Rust/someone that just wants to read the algorithm itself. What do you think?

---
> I hope one can squash commits in the PR.

Yeah, no worries, I always squash before merging and give it an appropriate name 👍🏻

**Comment by CodingTil** [comment]
> Could you elaborate on why you didn't just use one Sorter trait?

I think in the beginning I wanted to avoid having to require the trait `Copy` (or `Clone`) for all the `T`'s. So, if I could find a way to do without both, or just using `Clone`, I would've chosen that approach.
But I guess I have often simply called `sort_inplace` from `sort` anyway, thus requiring `Copy` almost everywhere.

Let me know if I should change it to your proposal. I do see that yours is a lot easier to read.

---
> I think I would rather have the trait implementation after the sorting function itself. That is, if you were to read top-to-bottom, you would find the algorithm first, and then the abstracting traits. This appeals better to someone new to Rust/someone that just wants to read the algorithm itself. What do you think?

Yeah, that sounds like a good idea. Will do!

**Comment by CodingTil** [comment]
> I think in the beginning I wanted to avoid having to require the trait Copy (or Clone) for all the T's. So, if I could find a way to do without both, or just using Clone, I would've chosen that approach.
But I guess I have often simply called sort_inplace from sort anyway, thus requiring Copy almost everywhere.
Let me know if I should change it to your proposal. I do see that yours is a lot easier to read.

Ah, what the heck. I am basically using Copy already everywhere. I'll use your proposal!

**Comment by CodingTil** [comment]
Alright, applied the suggestions. Requesting review :)

**Review by alexfertel — APPROVED** [review]
Awesome! Excellent PR tbh, I see you did quite a few small things in addition like removing some casts and removing some unnecessary trait bounds. Thank you!

**Pull request merged**